test/units/galaxy/test_api.py

```diff
@@ -974,6 +1028,89 @@ def test_cache_invalid_cache_content(content, cache_dir):
     assert stat.S_IMODE(os.stat(cache_file).st_mode) == 0o664
 
 
+def test_cache_complete_pagination(cache_dir, monkeypatch):
+
+    responses = get_collection_versions()
+    cache_file = os.path.join(cache_dir, 'api.json')
+
+    api = get_test_galaxy_api('https://galaxy.server.com/api/', 'v2', no_cache=False)
+
+    mock_open = MagicMock(
+        side_effect=[
+            StringIO(to_text(json.dumps(r)))
+            for r in responses
+        ]
+    )
+    monkeypatch.setattr(galaxy_api, 'open_url', mock_open)
+
+    actual_versions = api.get_collection_versions('namespace', 'collection')
+
+    with open(cache_file) as fd:
+        final_cache = json.loads(fd.read())
+        cached_server = final_cache['galaxy.server.com:']
+        cached_collection = cached_server['/api/v2/collections/namespace/collection/versions/']
+        cached_versions = [r['version'] for r in cached_collection['results']]
+
+    assert final_cache == api._cache
+    assert cached_versions == actual_versions == [u'1.0.0', u'1.0.1', u'1.0.2', u'1.0.3', u'1.0.4', u'1.0.5']
+
+
+def test_cache_flaky_pagination(cache_dir, monkeypatch):
+
+    responses = get_collection_versions()
+    cache_file = os.path.join(cache_dir, 'api.json')
+
+    api = get_test_galaxy_api('https://galaxy.server.com/api/', 'v2', no_cache=False)
+
+    # First attempt, fail midway through
+    mock_open = MagicMock(
+        side_effect=[
+            StringIO(to_text(json.dumps(responses[0]))),
+            StringIO(to_text(json.dumps(responses[1]))),
+            urllib_error.HTTPError(responses[1]['next'], 500, 'Error', {}, StringIO()),
+            StringIO(to_text(json.dumps(responses[3]))),
+        ]
+    )
+    monkeypatch.setattr(galaxy_api, 'open_url', mock_open)
+
+    with pytest.raises(GalaxyError):
+        actual_versions = api.get_collection_versions('namespace', 'collection')
+
+    with open(cache_file) as fd:
+        final_cache = json.loads(fd.read())
+
+    assert final_cache == {
+        'version': 1,
+        'galaxy.server.com:': {
+            'modified': {
+                'namespace.collection': responses[0]['modified']
+            }
+        }
+    }
+
+    # Reset API
+    api = get_test_galaxy_api('https://galaxy.server.com/api/', 'v2', no_cache=False)
+
+    # Second attempt is successful so cache should be populated
+    mock_open = MagicMock(
+        side_effect=[
+            StringIO(to_text(json.dumps(r)))
+            for r in responses
+        ]
+    )
+    monkeypatch.setattr(galaxy_api, 'open_url', mock_open)
+
+    actual_versions = api.get_collection_versions('namespace', 'collection')
```

Comment:
I think it may be cleaner to check this right away
```suggestion
    actual_versions = api.get_collection_versions('namespace', 'collection')
    assert actual_versions == [u'1.0.0', u'1.0.1', u'1.0.2', u'1.0.3', u'1.0.4', u'1.0.5']
```
and this would simplify the check at the end of this test.

Reply:
Thanks for the review!